src/compiler/checker.ts

```diff
@@ -14250,16 +14250,23 @@ namespace ts {
         }
 
         // Ordinarily we reduce a keyof M, where M is a mapped type { [P in K as N<P>]: X }, to simply N<K>. This however presumes
-        // that N distributes over union types, i.e. that N<A | B | C> is equivalent to N<A> | N<B> | N<C>. That presumption may not
-        // be true when N is a non-distributive conditional type or an instantiable type with a non-distributive conditional type as
-        // a constituent. In those cases, we cannot reduce keyof M and need to preserve it as is.
-        function maybeNonDistributiveNameType(type: Type | undefined): boolean {
-            return !!(type && (
-                type.flags & TypeFlags.Conditional && (!(<ConditionalType>type).root.isDistributive || maybeNonDistributiveNameType((<ConditionalType>type).checkType)) ||
-                type.flags & (TypeFlags.UnionOrIntersection | TypeFlags.TemplateLiteral) && some((<UnionOrIntersectionType | TemplateLiteralType>type).types, maybeNonDistributiveNameType) ||
-                type.flags & (TypeFlags.Index | TypeFlags.StringMapping) && maybeNonDistributiveNameType((<IndexType | StringMappingType>type).type) ||
-                type.flags & TypeFlags.IndexedAccess && maybeNonDistributiveNameType((<IndexedAccessType>type).indexType) ||
-                type.flags & TypeFlags.Substitution && maybeNonDistributiveNameType((<SubstitutionType>type).substitute)));
+        // that N distributes over union types, i.e. that N<A | B | C> is equivalent to N<A> | N<B> | N<C>. Specifically, we only
+        // want to perform the reduction when the name type of a mapped type is distributive with respect to the type variable
+        // introduced by the 'in' clause of the mapped type. Note that non-generic types are considered to be distributive because
+        // they're the same type regardless of what's being distributed over.
+        function hasDistributiveNameType(mappedType: MappedType) {
+            const typeVariable = getTypeParameterFromMappedType(mappedType);
+            return isDistributiveForTypeParameter(getNameTypeFromMappedType(mappedType) || typeVariable);
+            function isDistributiveForTypeParameter(type: Type): boolean {
+                return type.flags & TypeFlags.TypeParameter ? type === typeVariable :

```

Comment:
Not a bug per se, but isn't this check too strict? eg it causes

```
type NotDistributive<T, U, V = string> = keyof { [ K in keyof T as V & (K extends U ? K : never)]: T[K]};
```

to fail the `isDistributiveForTypeParameter`, because `V` is rejected by the `every` call.

Maybe add a flag to `isDistributiveForTypeParameter` saying whether or not you're under a checkType, default it to false (for the initial call, and the `every` call), pass true for the checkType call, and pass it through for the other cases?

Reply:
Agreed, fixed in the latest commit.